File: cmd_executor.c
#include "cmd_executor.h"
#include "cmd.h"
#include "cmd_parser.h"
#include "utils.h"
#include <fcntl.h>
#include <setjmp.h>
#include <stdio.h>
#include <sys/stat.h>
#include <unistd.h>

static char *cmd_executor_word_to_str(cmd_executor *executor, cmd *c,
                                      cmd_word *word);

static void cmd_executor_error(cmd_executor *executor, int status) {
  longjmp(executor->err_jmp, status);
}

cmd_executor *cmd_executor_new() {
  cmd_executor *executor = malloc(sizeof(cmd_executor));
  executor->vars = g_hash_table_new(g_str_hash, g_str_equal);
  executor->stdin_fno = STDIN_FILENO;
  executor->stdout_fno = STDOUT_FILENO;

  return executor;
}

static char *cmd_executor_get_var(cmd_executor *executor,
                                  cmd_word_part_var *var) {
  // Check if we have a var def for the command.
  char *var_val = g_hash_table_lookup(executor->vars, var->name->str);
  if (var_val != NULL) {
    return var_val;
  }

  // Fallback to the environment.
  return getenv(var->name->str);
}

static void set_var(cmd_executor *executor, cmd *c, cmd_var_assign *var,
                    GHashTable *vars) {
  char *value = cmd_executor_word_to_str(executor, c, var->value);

  // Copy the name and value so they don't reference memory owned by the
  // cmd (which will be freed later).
  char *name_cpy = malloc((strlen(var->name) + 1) * sizeof(char));
  char *value_cpy = malloc((strlen(value) + 1) * sizeof(char));
  strcpy(name_cpy, var->name);
  strcpy(value_cpy, value);

  g_hash_table_insert(vars, name_cpy, value_cpy);
}

static char *cmd_executor_word_to_str(cmd_executor *executor, cmd *c,
                                      cmd_word *word) {
  GString *res = g_string_new(NULL);

  for (GList *node = word->parts; node != NULL; node = node->next) {
    cmd_word_part *part = (cmd_word_part *)node->data;

    switch (part->type) {
    case CMD_WORD_PART_TYPE_LIT: {
      res = g_string_append(res, part->value.literal->str);
      break;
    }

    case CMD_WORD_PART_TYPE_STR: {
      cmd_word_part_str *str = part->value.str;
      if (str->quoted) {
        for (GList *str_node = str->parts; str_node != NULL;
             str_node = str_node->next) {
          cmd_word_part_str_part *str_part = str_node->data;

          switch (str_part->type) {
          case CMD_WORD_PART_STR_PART_TYPE_LITERAL: {
            res = g_string_append(res, str_part->value.literal->str);
            break;
          }

          case CMD_WORD_PART_STR_PART_TYPE_VAR: {
            char *val = cmd_executor_get_var(executor, str_part->value.var);
            if (val != NULL) {
              res = g_string_append(res, val);
            }

            break;
          }

          default:
            giveup("str part type not implemented");
          }
        }
      } else {
        for (GList *str_node = str->parts; str_node != NULL;
             str_node = str_node->next) {
          cmd_word_part_str_part *str_part = str_node->data;

          switch (str_part->type) {
          case CMD_WORD_PART_STR_PART_TYPE_LITERAL: {
            res = g_string_append(res, str_part->value.literal->str);
            break;
          }

          case CMD_WORD_PART_STR_PART_TYPE_VAR:
            giveup("cmd_executor_word_to_str: unexpected var");

          default:
            giveup("str part type not implemented");
          }
        }
      }

      break;
    }

    case CMD_WORD_PART_TYPE_VAR: {
      char *val = cmd_executor_get_var(executor, part->value.var);
      if (val != NULL) {
        res = g_string_append(res, val);
      }

      break;
    }

    case CMD_WORD_PART_TYPE_CMD_SUB: {
      // Keep track of the original fnos.
      int original_fnos[2] = {executor->stdin_fno, executor->stdout_fno};

      // Create a pipe.
      int pipe_fnos[2];
      if (pipe(pipe_fnos) < 0) {
        giveup("cmd_executor_exec: pipe failed");
      }

      int status;

      // Write to the write end of the pipe.
      executor->stdout_fno = pipe_fnos[1];
      status = cmd_executor_exec(executor, part->value.cmd_sub);

      // Restore the executor's stdout and close the write end of the pipe.
      executor->stdout_fno = original_fnos[1];
      close(pipe_fnos[1]);

      if (status != 0) {
        cmd_executor_error(executor, status);
      }

      // Read from the read end of the pipe.
      GString *arg = g_string_new(NULL);
      {
        char buf[BUFSIZ];

        ssize_t n;
        while ((n = read(pipe_fnos[0], buf, BUFSIZ)) != 0) {
          // Remove trailing newline.
          if (n < BUFSIZ) {
            buf[n - 1] = 0;
          }

          arg = g_string_append(arg, buf);
        }
      }

      // Close the read end of the pipe.
      close(pipe_fnos[0]);

      res = g_string_append(res, arg->str);

      break;
    }

    case CMD_WORD_PART_TYPE_PROC_SUB: {
      char file_name[] = "/tmp/turtle-proc-XXXXXX";
      int fd;
      if ((fd = mkstemp(file_name)) < 0) {
        giveup("cmd_executor_exec: proc sub file creation failed");
      }

      // Give permission rwx permissions to everyone.
      if (chmod(file_name, 0777) < 0) {
        giveup("cmd_executor_exec: proc sub file chmod failed");
      }

      int original_out_fd = executor->stdout_fno;

      // Redirect executor output to the write end of the pipe.
      executor->stdout_fno = fd;

      // Execute the command.
      int status;
      if ((status = cmd_executor_exec(executor, part->value.proc_sub)) != 0) {
        close(fd);

        executor->stdout_fno = original_out_fd;

        cmd_executor_error(executor, status);
      }

      // Close the file.
      close(fd);

      // Reopen the file with the correct flags.
      if ((fd = open(file_name, O_RDONLY)) < 0) {
        giveup("cmd_executor_exec: proc_sub reopen failed");
      }

      // Restore the original stdout fd.
      executor->stdout_fno = original_out_fd;

      // Append "/dev/fd/{pipe_read_end_fd}"
      g_string_append_printf(res, "%s", file_name);

      break;
    }

    default:
      giveup("cmd_executor_exec: unimplemented cmd_word_part_type");
    }
  }

  return res->str;
}

int cmd_executor_exec_term(cmd_executor *executor, char *term, char **argv) {
  if (strcmp(term, ".") == 0) {
    argv++;

    return cmd_executor_exec_term(executor, argv[0], argv);
  }

  pid_t pid;
  if ((pid = fork()) == 0) {
    if (executor->stdin_fno != STDIN_FILENO) {
      close(STDIN_FILENO);
      dup(executor->stdin_fno);
    }

    if (executor->stdout_fno != STDOUT_FILENO) {
      close(STDOUT_FILENO);
      dup(executor->stdout_fno);
    }

    execvp(term, argv);
    giveup("cmd_executor_exec_term: exec '%s' failed", term);
    exit(1);
  }

  int status;
  if ((pid = waitpid(pid, &status, 0)) < 0) {
    giveup("cmd_executor_exec_term: waitpid failed with pid=%d,status=%d", pid,
           status);
  }

  return status;
}

int cmd_executor_exec(cmd_executor *executor, cmd *cmd) {
  char *term = NULL;

  int argc = 0;
  GList *gargs = NULL;

  // Set up executor err jump.
  int err_status;
  if ((err_status = setjmp(executor->err_jmp)) != 0) {
    return err_status;
  }

  for (GList *node = cmd->parts; node != NULL; node = node->next) {
    cmd_part *part = (cmd_part *)node->data;

    switch (part->type) {
    case CMD_PART_TYPE_VAR_ASSIGN: {
      cmd_var_assign *var = part->value.var_assign;

      // If this is the only part of the command, set the var as an executor
      // var.
      if (node->next == NULL) {
        set_var(executor, cmd, var, executor->vars);
      }
      // Otherwise, set it as a var for the environment for the command.
      else {
        set_var(executor, cmd, var, cmd->env_vars);
      }

      break;
    }

    case CMD_PART_TYPE_WORD: {
      char *word = cmd_executor_word_to_str(executor, cmd, part->value.word);
      if (term == NULL) {
        argc++;

        term = word;
        gargs = g_list_append(gargs, term);
      } else {
        argc++;

        gargs = g_list_append(gargs, word);
      }

      break;
    }

    case CMD_PART_TYPE_PIPE: {
      int status;

      // Keep track of the original fnos.
      int original_fnos[2] = {executor->stdin_fno, executor->stdout_fno};

      // Create a pipe.
      int pipe_fnos[2];
      if (pipe(pipe_fnos) < 0) {
        giveup("cmd_executor_exec: pipe failed");
      }

      // Have the executor write to output of the pipe.
      executor->stdout_fno = pipe_fnos[1];

      // Execute the cmd we built.
      if ((status = cmd_executor_exec_term(
               executor, term, g_list_charptr_to_argv(gargs, argc))) != 0) {
        close(pipe_fnos[0]);
        close(pipe_fnos[1]);

        executor->stdin_fno = original_fnos[0];
        executor->stdout_fno = original_fnos[1];

        return status;
      }

      // Close the pipe output.
      close(pipe_fnos[1]);

      // Have the executor read from the input of the pipe and write to the
      // original out.
      executor->stdin_fno = pipe_fnos[0];
      executor->stdout_fno = original_fnos[1];

      // Execute the piped cmd.
      status = cmd_executor_exec(executor, part->value.piped_cmd);

      // Close the pipe input.
      close(pipe_fnos[0]);

      // Reset the executor stdin.
      executor->stdin_fno = original_fnos[0];

      return status;
    }

    case CMD_PART_TYPE_OR: {
      // Execute the command as-is.
      char **argv = g_list_charptr_to_argv(gargs, argc);
      int status = cmd_executor_exec_term(executor, term, argv);

      // If the result is 0, we're done!
      if (status == 0) {
        return status;
      }

      // Otherwise, execute the or'd command.
      return cmd_executor_exec(executor, part->value.or_cmd);
    }

    case CMD_PART_TYPE_AND: {
      // Execute the command as-is.
      char **argv = g_list_charptr_to_argv(gargs, argc);
      int status = cmd_executor_exec_term(executor, term, argv);

      // If the result non-zero, bail!
      if (status != 0) {
        return status;
      }

      // Otherwise, execute the and'd command.
      return cmd_executor_exec(executor, part->value.and_cmd);
    }

    default:
      giveup("cmd_executor_exec: not implemented");
    }
  }

  if (term == NULL) {
    return 0;
  }

  char **argv = g_list_charptr_to_argv(gargs, argc);
  return cmd_executor_exec_term(executor, term, argv);
}

File: src/executor/cmd_executor.zig
const std = @import("std");
const assert = std.debug.assert;

const cmd = @import("../parser/cmd.zig");

const proc = @import("proc.zig");
const util = @import("util.zig");
const c = util.c;

pub const ExecOpts = struct {
    stdin_fno: c_int,
    stdout_fno: c_int,
    wait: bool,
};

const ExecBuiltCmdResult = union(enum) {
    pid: i32,
    status: u8,
};

const ExecutableCmd = union(enum) {
    const Normal = struct {
        args: std.ArrayList([]u8),
        env: std.ArrayList([]u8),
    };

    const Branch = struct {
        left: ExecutableCmd,
        right: ExecutableCmd,
    };

    const Pipeline = struct {
        cmds: std.ArrayList(ExecutableCmd),
    };

    normal_cmd: Normal,

    or_cmd: *Branch,
    and_cmd: *Branch,

    pipeline: Pipeline,
};

pub const CmdExecutor = struct {
    const Self = @This();

    allocator: std.mem.Allocator,

    vars: std.StringHashMap([]u8),
    last_pid: ?c_int,
    last_status_code: ?u8,

    err_jmp_buf: ?c.jmp_buf,

    pub fn init(allocator: std.mem.Allocator) Self {
        return .{
            .allocator = allocator,
            .last_pid = null,
            .last_status_code = null,
            .vars = std.StringHashMap([]u8).init(allocator),
            .err_jmp_buf = null,
        };
    }

    pub fn exec(self: *Self, command: *cmd.Cmd, opts: ExecOpts) anyerror!u8 {
        self.err_jmp_buf = [_]c_int{0} ** 48;
        if (c.setjmp(&self.err_jmp_buf.?) != 0) {
            return self.last_status_code.?;
        }

        const built_cmd = try self.buildExecutableCmd(command);
        const result = try self.execBuiltCmd(built_cmd, opts);

        self.last_status_code = result.status;
        return self.last_status_code.?;
    }

    // TODO: we should really wait until the last second to perform cmd/proc subs! This is a bit surprising because building a cmd ends up having side effects!
    fn buildExecutableCmd(self: *Self, command: *cmd.Cmd) anyerror!ExecutableCmd {
        var args = std.ArrayList([]u8).init(self.allocator);
        const env = std.ArrayList([]u8).init(self.allocator);

        // Set up executor err jump.
        for (command.parts.items, 0..) |part, i| {
            switch (part) {
                .var_assign => |var_assign| {
                    const name = var_assign.name;
                    const value = try self.wordToStr(var_assign.value);

                    // If this is the only part of the command, set the var as an executor
                    // var.
                    if (i == command.parts.items.len - 1) {
                        try self.vars.put(name, value);
                    }
                    // Otherwise, set it as a var for the environment for the command.
                    else {
                        try command.env_vars.put(name, value);
                    }
                },

                .word => |word| {
                    try args.append(try self.wordToStr(word));
                },

                .pipeline => |pipeline_cmds| {
                    // TODO: check if there are any args/env/vars/etc.; that's probably an error with parsing if so!

                    var pipeline = try std.ArrayList(ExecutableCmd).initCapacity(self.allocator, pipeline_cmds.items.len);
                    for (pipeline_cmds.items) |pipeline_cmd| {
                        try pipeline.append(try self.buildExecutableCmd(pipeline_cmd));
                    }

                    return .{
                        .pipeline = .{
                            .cmds = pipeline,
                        },
                    };
                },

                .or_cmd => |or_cmd| {
                    const built_or_cmd = try self.allocator.create(ExecutableCmd.Branch);
                    built_or_cmd.* = .{
                        .left = .{
                            .normal_cmd = .{
                                .args = args,
                                .env = env,
                            },
                        },
                        .right = try self.buildExecutableCmd(or_cmd),
                    };

                    return .{ .or_cmd = built_or_cmd };
                },

                .and_cmd => |and_cmd| {
                    const built_and_cmd = try self.allocator.create(ExecutableCmd.Branch);
                    built_and_cmd.* = .{
                        .left = .{
                            .normal_cmd = .{
                                .args = args,
                                .env = env,
                            },
                        },
                        .right = try self.buildExecutableCmd(and_cmd),
                    };

                    return .{ .and_cmd = built_and_cmd };
                },
            }
        }

        return .{
            .normal_cmd = .{
                .args = args,
                .env = env,
            },
        };
    }

    fn wordToStr(self: *Self, word: *cmd.CmdWord) anyerror![]u8 {
        var res = std.ArrayList(u8).init(self.allocator);

        for (word.parts.items) |part| {
            switch (part.*) {
                .literal => |literal| {
                    try res.appendSlice(literal);
                },
                .str => |str| {
                    if (str.expandable) {
                        for (str.parts.items) |str_part| {
                            switch (str_part) {
                                .literal => |literal| {
                                    try res.appendSlice(literal);
                                },

                                .variable => |variable| {
                                    const val = try self.getVar(variable.name);
                                    if (val != null) {
                                        try res.appendSlice(val.?);
                                    }
                                },
                            }
                        }
                    } else {
                        for (str.parts.items) |str_part| {
                            switch (str_part) {
                                .literal => |literal| {
                                    try res.appendSlice(literal);
                                },

                                .variable => |_| {
                                    @panic("not implemented");
                                },
                            }
                        }
                    }
                },

                .variable => |variable| {
                    const val = try self.getVar(variable.name);
                    if (val != null) {
                        try res.appendSlice(val.?);
                    }
                },

                .cmd_sub => |cmd_sub| {
                    var pipe_fnos = [2]c_int{ 0, 0 };
                    if (c.pipe(&pipe_fnos) < 0) {
                        util.giveup("wordToStr: pipe failed", .{});
                    }

                    // Write to the write end of the pipe while executing the cmd.
                    const status = try self.exec(cmd_sub, .{
                        .stdin_fno = std.posix.STDIN_FILENO,
                        .stdout_fno = pipe_fnos[1],
                        .wait = true,
                    });

                    // Signal that we're done writing.
                    _ = c.close(pipe_fnos[1]);

                    // If the command failed, bail!
                    if (status != 0) {
                        self.exitErr(status);
                    }

                    // Read the result.
                    var arg = blk: {
                        var argRes = std.ArrayList(u8).init(self.allocator);

                        var buf = [_]u8{0} ** c.BUFSIZ;

                        var n: usize = @intCast(c.read(pipe_fnos[0], &buf, buf.len));
                        while (n != 0) {
                            var bufSlice: []u8 = buf[0..];

                            if (n < c.BUFSIZ) {
                                // Remove the trailing newline.
                                bufSlice = bufSlice[0..n];
                                bufSlice[n - 1] = 0;
                            }

                            try argRes.appendSlice(bufSlice);

                            n = @intCast(c.read(pipe_fnos[0], &buf, buf.len));
                        }

                        break :blk argRes;
                    };

                    // Close the read end.
                    _ = c.close(pipe_fnos[0]);

                    try res.appendSlice(try arg.toOwnedSlice());
                },

                .proc_sub => |proc_sub| {
                    const file_name_template = "/tmp/turtle-proc-XXXXXX\x00";
                    var file_name_buf: [file_name_template.len:0]u8 = undefined;
                    std.mem.copyForwards(u8, &file_name_buf, file_name_template);

                    var fd = c.mkstemp(file_name_buf[0..].ptr);
                    if (fd < 0) {
                        util.giveup("wordToStr: proc sub file creation failed", .{});
                    }

                    if (c.chmod(file_name_buf[0..].ptr, 0o777) < 0) {
                        util.giveup("wordToStr: proc sub file chmod failed", .{});
                    }

                    // Write to the file during execution.
                    const maybeStatus = self.exec(proc_sub, .{
                        .stdin_fno = std.posix.STDIN_FILENO,
                        .stdout_fno = fd,
                        .wait = true,
                    });

                    // Signal that we're done writing to the file.
                    _ = c.close(fd);

                    // If the command failed, bail!
                    const status = try maybeStatus;
                    if (status != 0) {
                        self.exitErr(status);
                    }

                    // Reopen the file with the correct flags.
                    fd = c.open(file_name_buf[0..].ptr, c.O_RDONLY);
                    if (fd < 0) {
                        util.giveup("wordToStr: proc sub file reopen failed", .{});
                    }

                    // Provide the filename as "/dev/fd/$FD".
                    try res.appendSlice(try std.fmt.allocPrint(self.allocator, "/dev/fd/{d}", .{fd}));
                },
            }
        }

        return res.toOwnedSlice();
    }

    fn getVar(self: Self, name: []u8) !?[]u8 {
        // Check if this is a special var name.
        if (std.mem.eql(u8, "!", name)) {
            return try std.fmt.allocPrint(self.allocator, "{d}", .{self.last_pid.?});
        }
        if (std.mem.eql(u8, "?", name)) {
            return try std.fmt.allocPrint(self.allocator, "{d}", .{self.last_status_code.?});
        }

        // Check if we have a var def for the command.
        var value = self.vars.get(name);
        if (value == null) {
            // Fall back to the environment.
            const envVal = std.posix.getenv(name);
            if (envVal != null) {
                const buf = try self.allocator.alloc(u8, envVal.?.len);
                @memcpy(buf, envVal.?[0..envVal.?.len]);

                value = buf;
            }
        }

        return value;
    }

    // TODO: clean up memory once we're done.
    fn execBuiltCmd(self: *Self, command: ExecutableCmd, opts: ExecOpts) anyerror!ExecBuiltCmdResult {
        switch (command) {
            .normal_cmd => |normal_cmd| {
                if (normal_cmd.args.items.len == 0) {
                    return .{ .status = 0 };
                }

                const fork_exec_args = proc.ForkExecArgs{
                    .argv = normal_cmd.args.items,
                    .envp = normal_cmd.env.items,
                    .stdin_fno = opts.stdin_fno,
                    .stdout_fno = opts.stdout_fno,
                };

                if (opts.wait) {
                    return .{ .status = try proc.forkExec(self.allocator, fork_exec_args) };
                } else {
                    return .{ .pid = try proc.forkExecNoWait(self.allocator, fork_exec_args) };
                }
            },
            .or_cmd => |or_cmd| {
                const left_status = try self.execBuiltCmd(or_cmd.left, opts);

                // If the left side succeeded, we're done!
                if (left_status.status == 0) {
                    return left_status;
                }

                // Otherwise, execute the or'd commnd.
                return try self.execBuiltCmd(or_cmd.right, opts);
            },
            .and_cmd => |or_cmd| {
                const left_status = try self.execBuiltCmd(or_cmd.left, opts);

                // If the left side failed, we're done!
                if (left_status.status != 0) {
                    return left_status;
                }

                // Otherwise, keep going!
                return try self.execBuiltCmd(or_cmd.right, opts);
            },
            .pipeline => |pipeline| {
                return try self.runPipeline(pipeline, opts);
            },
        }
    }

    fn runPipeline(self: *Self, pipeline: ExecutableCmd.Pipeline, opts: ExecOpts) !ExecBuiltCmdResult {
        const pid = c.fork();
        assert(pid != -1);
        if (pid == 0) {
            _ = c.setpgid(0, 0);

            const PipelineProcInfo = struct {
                pid: i32,
                stdin_fno: c_int,
                stdout_fno: c_int,
            };

            var pipeline_procs = std.ArrayList(PipelineProcInfo).init(self.allocator);

            // Start up each process in the pipeline.
            var prev_stdin_fno = opts.stdin_fno;
            for (pipeline.cmds.items, 0..) |pipeline_cmd, i| {
                const is_last_pipeline_proc = i == pipeline.cmds.items.len - 1;

                // TODO: make sure we don't have any in-progress args/env/etc. because that would indicate an error with the parsing (since the pipeline cmds should be self-contained).

                // Figure out what our fnos should be.
                //
                // If this is the last command in the pipeline, use the previous stdin fno, but output directly to stdout.
                // Otherwise, allocate a pipe we'll use to pipe output between procs.
                var fnos: [2]c_int = undefined;
                if (is_last_pipeline_proc) {
                    fnos = .{ prev_stdin_fno, opts.stdout_fno };
                } else {
                    var pipe_fnos = [2]c_int{ 0, 0 };
                    _ = c.pipe(&pipe_fnos);

                    fnos = .{ prev_stdin_fno, pipe_fnos[1] };

                    // Save the read end for the next pipeline proc.
                    prev_stdin_fno = pipe_fnos[0];
                }

                // Fork-exec the left side but don't wait for it.
                const exec_result = try self.execBuiltCmd(pipeline_cmd, .{
                    .stdin_fno = fnos[0],
                    .stdout_fno = fnos[1],
                    .wait = false,
                });

                _ = c.close(fnos[0]);
                _ = c.close(fnos[1]);

                // Add the left side of the pipe to the pipeline.
                try pipeline_procs.append(.{
                    .pid = exec_result.pid,
                    .stdin_fno = fnos[0],
                    .stdout_fno = fnos[1],
                });
            }

            // Close the last stdin fno.
            _ = c.close(prev_stdin_fno);

            var exit_status: u8 = 0;
            for (pipeline_procs.items, 0..) |pipeline_proc, i| {
                const status = proc.wait(pipeline_proc.pid);

                // If this proc failed, kill the rest of the procs in the pipeline and bail.
                if (status != 0) {
                    for (pipeline_procs.items[i + 1 ..]) |next_cmd| {
                        _ = c.kill(next_cmd.pid, c.SIGKILL);
                    }

                    exit_status = status;
                    break;
                }
            }

            std.posix.exit(exit_status);
        }

        const status = proc.wait(pid);
        return .{
            .status = status,
        };
    }

    fn exitErr(self: *Self, status: u8) noreturn {
        self.last_status_code = status;
        c.longjmp(&self.err_jmp_buf.?, status);
        unreachable;
    }
};
